src/hotspot/share/utilities/zipLibrary.cpp

```diff
@@ -0,0 +1,186 @@
+  assert(_zip_handle != nullptr, "invariant");
+  void* func = os::dll_lookup(_zip_handle, name);
+  if (func == nullptr) {
+    char msg[256] = "";
+    jio_snprintf(&msg[0], sizeof msg, "Could not resolve \"%s\"", name);
+    vm_exit_during_initialization(&msg[0], path);
+  }
+  return func;
+}
+
+static void store_function_pointers(const char* path) {
+  assert(_zip_handle != nullptr, "invariant");
+  ZIP_Open = CAST_TO_FN_PTR(ZIP_Open_t, dll_lookup("ZIP_Open", path));
+  ZIP_Close = CAST_TO_FN_PTR(ZIP_Close_t, dll_lookup("ZIP_Close", path));
+  ZIP_FindEntry = CAST_TO_FN_PTR(ZIP_FindEntry_t, dll_lookup("ZIP_FindEntry", path));
+  ZIP_ReadEntry = CAST_TO_FN_PTR(ZIP_ReadEntry_t, dll_lookup("ZIP_ReadEntry", path));
+  ZIP_CRC32 = CAST_TO_FN_PTR(ZIP_CRC32_t, dll_lookup("ZIP_CRC32", path));
+  ZIP_GZip_InitParams = CAST_TO_FN_PTR(ZIP_GZip_InitParams_t, dll_lookup("ZIP_GZip_InitParams", path));
+  ZIP_GZip_Fully = CAST_TO_FN_PTR(ZIP_GZip_Fully_t, dll_lookup("ZIP_GZip_Fully", path));
+}
+
+static void load_zip_library() {
+  assert(!is_loaded(), "should not load zip library twice");
+  char path[JVM_MAXPATHLEN];
+  if (os::dll_locate_lib(&path[0], sizeof path, Arguments::get_dll_dir(), "zip")) {
+    char ebuf[1024];
+    _zip_handle = os::dll_load(&path[0], &ebuf[0], sizeof ebuf);
+  }
+  if (_zip_handle == nullptr) {
+    vm_exit_during_initialization("Unable to load zip library", &path[0]);
+  }
+  store_function_pointers(&path[0]);
+  Atomic::release_store(&_loaded, true);
+  assert(is_loaded(), "invariant");
+}
+
+//
+// Helper mutex class that also ensures that java threads
+// are in _thread_in_native when loading the zip library.
+//
+class ZipLibraryLoaderLock : public StackObj {
+ private:
+  static Semaphore _lock;
+  JavaThread* _jt;
+ public:
+   ZipLibraryLoaderLock() : _jt(nullptr) {
+    Thread* thread = Thread::current_or_null();
+    if (thread != nullptr && thread->is_Java_thread()) {
+      JavaThread* const jt = JavaThread::cast(thread);
+      if (jt->thread_state() != _thread_in_native) {
+        _jt = jt;
+        ThreadStateTransition::transition_from_vm(jt, _thread_in_native, false);
+      }
+    }
+    _lock.wait();
+  }
+  ~ZipLibraryLoaderLock() {
+    _lock.signal();
+    if (_jt != nullptr) {
+      ThreadStateTransition::transition_from_native(_jt, _thread_in_vm, false);
+    }
+  }
+};
+
+Semaphore ZipLibraryLoaderLock::_lock(1);
+
+static void initialize() {
+  if (is_loaded()) {
+    return;
+  }
+  ZipLibraryLoaderLock lock;
+  if (not_loaded()) {
+    load_zip_library();
+  }
+}
+
+void** ZipLibrary::open(const char* name, char** pmsg) {
+  initialize();
+  assert(ZIP_Open != nullptr, "invariant");
+  return ZIP_Open(name, pmsg);
+}
+
+void ZipLibrary::close(jzfile* zip) {
+  assert(is_loaded(), "invariant");
+  assert(ZIP_Close != nullptr, "invariant");
+  ZIP_Close(zip);
+}
+
+jzentry* ZipLibrary::find_entry(jzfile* zip, const char* name, jint* sizeP, jint* nameLen) {
+  initialize();
```

Comment:
Again don't we have to have initialized things to get the library we are now searching? The same goes for all(?) the other methods.

Reply:
No, this is lazy library initialization of the library as part of all the API entries (except close).

Reply:
ZipLibrary is a static facade exposing the entry points implemented in the "real" native zip library.

Reply:
Realized that the usage from HeapDumperCompression should not exit the VM if there is a problem loading up the zip library. Now updated to support that usage better.

Reply:
Okay so if we have no control over the order in which functions are called then they all must initialize, including `close` - no?

Reply:
Close seems obvious, as you pointed out earlier, to be subsequent from open. An assert is there now, to capture erroneous usages, which I think is the correct way to capture wrong callers.

The complication comes with init_params() and compress() which are used in a jcmd heap dump routine, it does not open anything, it merely wants to re-use the logic of functions in the zip library.

Reply:
Hmm okay. I found it hard to see exactly how this API actually gets exported for use.